### ISSUE TITLE
[Question] How can I intercept same target method multi times in single transformer ?

### ISSUE THREAD

**Issue by anylain** [raphw/byte-buddy#762]
Hi! I'm developing a APM system based on byte-buddy. sometimes I may have some different plugin to intercept same target method, and I find out in this situation only one plugin can work.

here is a demo

``` java
package com.example.demo;

import net.bytebuddy.agent.ByteBuddyAgent;
import net.bytebuddy.agent.builder.AgentBuilder;
import net.bytebuddy.description.type.TypeDescription;
import net.bytebuddy.dynamic.DynamicType;
import net.bytebuddy.implementation.MethodDelegation;
import net.bytebuddy.implementation.bind.annotation.*;
import net.bytebuddy.utility.JavaModule;

import java.lang.instrument.Instrumentation;
import java.lang.reflect.Method;
import java.util.concurrent.Callable;

import static net.bytebuddy.matcher.ElementMatchers.named;

public class ByteBuddyTest {

    public static void main(String[] args) {

        Instrumentation instrumentation = ByteBuddyAgent.install();
        new AgentBuilder.Default()
                .with(AgentBuilder.RedefinitionStrategy.REDEFINITION)
                .type(named("com.example.demo.ByteBuddyTest$TargetClass"))
                .transform(new AgentBuilder.Transformer() {
                    @Override
                    public DynamicType.Builder<?> transform(DynamicType.Builder<?> builder, TypeDescription typeDescription, ClassLoader classLoader, JavaModule module) {
                        return builder
                                .method(named("targetMethod"))
                                .intercept(MethodDelegation.to(new InvokerA()))
                                .method(named("targetMethod"))
                                .intercept(MethodDelegation.to(new InvokerB()));
                    }
                })
                .installOn(instrumentation);


        TargetClass targetClass = new TargetClass();
        targetClass.targetMethod();

    }


    public static class TargetClass {

        public void targetMethod() {
            System.out.println("targetMethod invoked");
        }

    }

    public static class InvokerA {
        @RuntimeType
        public Object intercept(@This Object obj,
                                @AllArguments Object[] allArguments,
                                @SuperCall Callable<?> origin,
                                @Origin Method method) throws Throwable {

            System.out.println(this.getClass().getSimpleName());

            return origin.call();
        }
    }

    public static class InvokerB {
        @RuntimeType
        public Object intercept(@This Object obj,
                                @AllArguments Object[] allArguments,
                                @SuperCall Callable<?> origin,
                                @Origin Method method) throws Throwable {

            System.out.println(this.getClass().getSimpleName());

            return origin.call();
        }
    }

}

```

The output I expected :
```
InvokerB
InvokerA
targetMethod invoked
```

The actual output :
```
InvokerB
targetMethod invoked
```

What should I do to get InvokerA and InvokerB both to work ?

PS: Thanks for the awesome work! byte-buddy has helped me a lot.

**Comment by raphw** [comment]
If you have plugins, you should probably work with decorative instrumentation, not with replacing method bodies. Have a look at `Advice` to add code at the beginning and end of a method what is combinable.

**Comment by anylain** [comment]
Is there a difference between `advice` and `replacing method bodies` in terms of performance?

**Comment by anylain** [comment]
Hi, I learned the documentation about `Advice` and found that this solution is not very suitable for me. In the APM system I developed, all plugins implement several interfaces to be hooked up with different modules in the system, so each plugin needs to be instantiated before use, but `Advice` can only use static methods. It made it difficult for me to combine it with the plugin instance.
Except for `Advice`, can I have other options?

**Comment by raphw** [comment]
In this case, you would have to merge the interceptors as in:

```java
MethodDelegation.to(new InvokerB())
  .andThen(MethodDelegation.to(new InvokerA())
  .andThen(SuperMethodCall.INSTANCE);
```

By the latter, you can also drop the `callSuper` part as it is explicit.

Simply add the above last and match it with a combination of both interceptors matchers.

As for advice, you can still use it but you would need to add your state to a (static) field that you add explicitly and access via `@Advice.FieldValue`.

**Comment by anylain** [comment]
Sorry, I described the problem a bit simple.
In fact, the scope of my different `invoker` work is not the same, for example, invokerA will handle methods with annotations of `@A`, but `InvokerB` may handle methods with `@B`, these annotations are possible In the same method, it may also be on different methods.

I have tried to initialize a separate `AgentBuilder` for each plugin, which can solve this problem, but there are a lot of plugins. I am worried that too many AgentBuilder will affect the performance of Class loading.

In addition, I also considered your suggestion, put the plugin in the Field and then use `@Advice.FieldValue` to access, but there will still be some problems, because I need to time the method call time, so I need a context to Passing between `methodEnter` and `methodExit`, I tried to solve this problem with `WithCustomMapping.bind`, but it didn't work.
